src/cpp/b2w.cpp

```diff
@@ -30,138 +30,78 @@ UNSW, Sydney Australia
+    const int i = pos - win_b;     // base position within window
+    char* mm = rd + i;             // location of base
+    if (pos >= win_b && pos <= win_e) {  // make sure read position within window
         if (pl->is_del == 1) {                 // remove deletions, replacing with reference
             // char *rb;
             // int td;
             // td = pl->b->core.tid;
             // rb = fai_fetch(tmp->fai, tmp->in->header->target_name[td], &tmp->len);
             *mm = '-';
-        } else {  // replace 'N' in tmp.read with current read base
-            *mm = bam_nt16_rev_table[bam1_seqi(bam1_seq(pl->b), pl->qpos)];
+        } else {  // replace 'N' in rd with current read base
+            // based on samtools/bam.h
+            *mm = seq_nt16_str[bam_seqi(bam_get_seq(pl->b), pl->qpos)];
         }
+        // NOTE samtools pileup mode without a ref behave this way, they do not ignore bases that are shifted by indel  e.g.: samtools 'C-6NNNNNN' is considered as 'C' by b2w
+        // NOTE fil (and samtools pileup mode with a ref) do not behaves exactly this way, they do ignore bases that are shifted by indel   e.g.: samtools 'C-6NNNNNN' is considered as '*' by fil. This can cause a BUG in fil.
     }
     return 0;
 }
 
-// callback two (for printing reads)
-static int fetch_func2(const bam1_t* b, void* data)
-{
-    tmpstruct_t* tmp = (tmpstruct_t*)data;
-    uint32_t Rstart, Rend;
-    Rstart = b->core.pos;
-    Rend = bam_calend(&b->core, bam1_cigar(b));
-    int* cv = (int*)tmp->rLen + Rstart;
-    *cv += 1;
-    int readLen = Rend - Rstart;
-    if ((*cv < tmp->max) && (readLen >= tmp->min_overlap)) {
-        char rd[readLen + 1];
-        for (int i = 0; i < readLen; i++) {
-            rd[i] = 'N';
-        }
-        rd[readLen] = 0;
-        tmp->rd = rd;
-        tmp->beg = Rstart;
-        tmp->end = Rend;
-        bam_plbuf_push(b, tmp->buf);
-        bam_plbuf_push(0, tmp->buf);
-        bam_plbuf_reset(tmp->buf);
-        tmp->reads << bam1_qname(b) << "\t" << tmp->b << "\t" << tmp->e << "\t" << Rstart + 1
-                   << "\t" << Rend << "\t" << rd << "\n";
-    }
-    return 0;
-}
 
 // main
 int main(int argc, char* argv[])
 {
     int c = 0;           // for parsing command line arguments
-    int ref = 0;         // index for retrieving reference chromosome
-    char filename[100];  // filename for window files
 
+    htsFile* inFile;
+    faidx_t* fai;
+    std::ofstream reads;
     std::ofstream covOut;  // stores window coverages
 
-    tmpstruct_t tmp;  // data for callback functions
+    paramstruct_t param = {  // data for callback functions
+        // based on default from cli.py and shotgun.py
+        .win = 201,   // window size
+        .inc = 201/3, // incr = size / shifts
+        .min_overlap = (int) std::round(201. * 0.85), // min_overlap = size * win_min_ext
+        .max = 10000 / 201 // max_c = max_coverage / size
+    };
```

Comment:
all of this can be done with simple aggregate initialization

Reply:
indeed, this is the only line with designated initializers. Enabling GNUisms over ISO C++ for one initialization will not fly.